@xen-orchestra/backups-cli/index.js

```diff
@@ -0,0 +1,354 @@
+    new Set(entries.filter(_ => _.endsWith('.xva'))),
+  ])
+
+  await asyncMap(xvas, async path => {
+    // check is not good enough to delete the file, the best we can do is report
+    // it
+    if (!(await isValidTar(path))) {
+      console.warn('Potential broken XVA', path)
+      console.warn('')
+    }
+  })
+
+  const unusedVhds = new Set(vhds)
+  const unusedXvas = new Set(xvas)
+
+  await asyncMap(jsons, async json => {
+    const metadata = JSON.parse(await fs.readFile(json))
+    const { mode } = metadata
+    if (mode === 'full') {
+      const linkedXva = resolve(vmDir, metadata.xva)
+
+      if (xvas.has(linkedXva)) {
+        unusedXvas.delete(linkedXva)
+      } else {
+        console.warn('Error while checking backup', json)
+        console.warn('  missing file', linkedXva)
+        force && console.warn('  deleting…')
+        console.warn('')
+        force && (await handler.unlink(json))
+      }
+    } else if (mode === 'delta') {
+      let linkedVhds = metadata.vhds
+      linkedVhds = Object.keys(linkedVhds).map(key =>
+        resolve(vmDir, linkedVhds[key])
+      )
+
+      if (linkedVhds.every(_ => vhds.has(_))) {
+        linkedVhds.forEach(_ => unusedVhds.delete(_))
+      } else {
+        console.warn('Error while checking backup', json)
+        linkedVhds
+          .filter(_ => !vhds.has(_))
+          .forEach(linkedVhd => {
+            console.warn('  missing file', linkedVhd)
+          })
+        force && console.warn('  deleting…')
+        console.warn('')
+        force && (await handler.unlink(json))
+      }
+    }
+  })
+
+  // TODO: parallelize by vm/job/vdi
+  const unusedVhdsDeletion = []
+  {
+    // VHD chains (as list from child to ancestor) to merge indexed by last
+    // ancestor
+    const vhdChainsToMerge = { __proto__: null }
+
+    const toCheck = new Set(unusedVhds)
+
+    const getUsedChildChainOrDelete = vhd => {
+      if (vhd in vhdChainsToMerge) {
+        const chain = vhdChainsToMerge[vhd]
+        delete vhdChainsToMerge[vhd]
+        return chain
+      }
+
+      if (!unusedVhds.has(vhd)) {
+        return [vhd]
+      }
+
+      // no longer needs to be checked
+      toCheck.delete(vhd)
+
+      const child = vhdChildren[vhd]
+      if (child !== undefined) {
+        const chain = getUsedChildChainOrDelete(child)
+        if (chain !== undefined) {
+          chain.push(vhd)
+          return chain
+        }
+      }
+
+      console.warn('Unused VHD', vhd)
+      force && console.warn('  deleting…')
+      console.warn('')
+      force && unusedVhdsDeletion.push(handler.unlink(vhd))
+    }
+
+    toCheck.forEach(vhd => {
+      vhdChainsToMerge[vhd] = getUsedChildChainOrDelete(vhd)
+    })
+
+    Object.keys(vhdChainsToMerge).forEach(key => {
+      const chain = vhdChainsToMerge[key]
+      if (chain !== undefined) {
+        unusedVhdsDeletion.push(mergeVhdChain(chain))
+      }
+    })
+  }
+
+  await Promise.all([
+    unusedVhdsDeletion,
```

Comment:
Don't the `mergeVhdChain` need to be `await`ed before calling the `unlink`s?

Reply:
Nope, because they are not touching the same files: on one hand, the unused VHDs are deleted, on the other hand, the chains are merged and then the unused VHDs are deleted.